File: packages/flutter-im-ui/test/image_group_message_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

/// The shared layout table (`spec/image-group-layout-vectors.json`) and the
/// album body built on it.
void main() {
  final table =
      jsonDecode(
            File(
              '../../spec/image-group-layout-vectors.json',
            ).readAsStringSync(),
          )
          as Map<String, dynamic>;

  group('image-group layout', () {
    test('draws at most the table\'s number of tiles', () {
      expect(flareImageGroupMaxVisible, table['maxVisible']);
    });
    for (final raw in (table['cases'] as List).cast<Map<String, dynamic>>()) {
      test('${raw['count']} image(s)', () {
        expect(
          flareImageGroupLayout(raw['count'] as int),
          FlareImageGroupLayout(
            columns: raw['columns'] as int,
            visible: raw['visible'] as int,
            more: raw['more'] as int,
          ),
        );
      });
    }
  });

  group('FlareImageGroupMessage', () {
    List<FlareImageContent> album(int count) => [
      for (var i = 0; i < count; i++) FlareImageContent(url: ''),
    ];

    Future<void> pump(WidgetTester tester, Widget child) => tester.pumpWidget(
      MaterialApp(
        home: Scaffold(body: Center(child: child)),
      ),
    );

    testWidgets('lays tiles out by the rule and covers the last', (
      tester,
    ) async {
      await pump(tester, FlareImageGroupMessage(images: album(12)));
      expect(find.text('+4'), findsOneWidget);
      // Nine tiles, three to a row: every tile is (240 - 2 gaps) / 3 wide.
      final tile = tester.getSize(find.bySemanticsLabel(RegExp('第 1 张图片')));
      expect(tile.width, closeTo((240 - 2 * FlareSizes.spacingXs) / 3, 0.01));
      expect(find.bySemanticsLabel(RegExp('第 9 张图片.*另有 4 张')), findsOneWidget);
      expect(find.bySemanticsLabel(RegExp('第 10 张图片')), findsNothing);

      await pump(tester, FlareImageGroupMessage(images: album(4)));
      expect(find.textContaining('+'), findsNothing);
      final square = tester.getSize(find.bySemanticsLabel(RegExp('第 1 张图片')));
      expect(square.width, closeTo((240 - FlareSizes.spacingXs) / 2, 0.01));
    });

    testWidgets('a tile opens its own image, the covered one included', (
      tester,
    ) async {
      final opened = <int>[];
      await pump(
        tester,
        FlareImageGroupMessage(
          images: album(12),
          description: '周末爬山',
          onOpen: opened.add,
        ),
      );
      await tester.tap(find.bySemanticsLabel(RegExp('第 2 张图片')));
      await tester.tap(find.text('+4'));
      expect(opened, [1, 8]);
      expect(find.text('周末爬山'), findsOneWidget);
      expect(find.bySemanticsLabel('12 张图片'), findsOneWidget);
    });

    testWidgets('an album without images draws nothing', (tester) async {
      await pump(tester, const FlareImageGroupMessage(images: []));
      expect(find.byType(Image), findsNothing);
      expect(tester.getSize(find.byType(FlareImageGroupMessage)), Size.zero);
    });

    testWidgets(
      'the dispatcher draws album content and hands the host the album',
      (tester) async {
        final content = FlareImageGroupContent(images: album(3));
        FlareMessageContent? handled;
        await pump(
          tester,
          FlareMessageContentView(
            content: content,
            onMediaAction: (value) => handled = value,
          ),
        );
        expect(find.byType(FlareImageGroupMessage), findsOneWidget);
        await tester.tap(find.bySemanticsLabel(RegExp('第 3 张图片')));
        expect(handled, same(content));
        expect(
          flareMessagePreviewText(content, const FlareStrings()),
          const FlareStrings().previewImageGroupCount(3),
        );
      },
    );
  });
}

File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/ImageGroupMessageTest.kt
package com.flare.im.ui

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/** The shared layout table (`spec/image-group-layout-vectors.json`) and what the album body builds on it. */
class ImageGroupMessageTest {
    private fun vectorsFile(): File {
        var dir: File? = File(".").absoluteFile
        while (dir != null) {
            val candidate = File(dir, "spec/image-group-layout-vectors.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile
        }
        error("spec/image-group-layout-vectors.json not found")
    }

    @Test fun everyCountLaysOutAsTheTableSays() {
        val table = FlareJson.parse(vectorsFile().readText()) as Map<*, *>
        assertEquals(FLARE_IMAGE_GROUP_MAX_VISIBLE, (table["maxVisible"] as Double).toInt())
        val cases = table["cases"] as List<*>
        assertTrue(cases.size >= 13)
        for (raw in cases) {
            val case = raw as Map<*, *>
            val count = (case["count"] as Double).toInt()
            assertEquals(
                FlareImageGroupLayout((case["columns"] as Double).toInt(), (case["visible"] as Double).toInt(), (case["more"] as Double).toInt()),
                flareImageGroupLayout(count),
                "$count image(s)",
            )
        }
    }

    @Test fun tilesAreNamedByPositionAndTheCoveredOneCountsWhatIsNotDrawn() {
        val strings = FlareStrings()
        val layout = flareImageGroupLayout(12)
        assertEquals("第 1 张图片，共 12 张", flareImageGroupTileLabel(0, 12, layout, strings))
        assertEquals("第 9 张图片，共 12 张，另有 4 张未显示", flareImageGroupTileLabel(8, 12, layout, strings))
        assertEquals(false, layout.covers(7))
    }

    @Test fun anAlbumTapGoesToTheHostOrPreviewsTheTileImage() {
        val album = FlareImageGroupContent(listOf(FlareImageContent("https://cdn/a.jpg"), FlareImageContent("https://cdn/b.jpg")))
        assertEquals(FlareContentTap.Host, flareContentTap(album, hasMediaHandler = true, hasFileHandler = false))
        assertEquals(FlareContentTap.PreviewAlbumImage, flareContentTap(album, hasMediaHandler = false, hasFileHandler = false))
    }

    @Test fun anAlbumSummarisesAsItsCount() {
        val strings = FlareStrings()
        assertEquals(strings.previewImageGroup, flareMessagePreviewText(FlareImageGroupContent(emptyList()), strings))
        assertEquals(strings.previewImageGroupCount(2), flareMessagePreviewText(FlareImageGroupContent(List(2) { FlareImageContent("") }), strings))
    }
}
